_posts/2024-09-24-from-jquery-to-hotwire-lessons-of-a-successful-migration.md

```diff
@@ -0,0 +1,361 @@
+
+### 4. Beware of the links
+
+So far I only gave examples of adding Turbo to new views. But its obtrusive nature really gets obvious when you start to add Turbo Frames to existing code. As in, there is a block of HTML that you wrap with a Turbo Frame because you want to do Turbo things in it.
+
+Well, you already know about the `DOMContentLoaded` part: if you have any JavaScript code listening to some element via their class name (id, etc) in that block of HTML, you'll have to migrate this JavaScript code to a Stimulus controller. This may or may not be straightforward to identify.
+
+But there is another thing I want to point out: by default, a link within a turbo frame targets _the frame itself_. What that means is, if you have some code like this:
+
+```html
+<!-- some HTML before the frame, for example: -->
+<img src="/hello.jpg" alt="hello" />
+
+<turbo-frame id="my-frame">
+  <a href="/resources/3">Show resource</a>
+</turbo-frame>
+
+<!-- more HTML here, for example: -->
+<img src="/bye.jpg" alt="bye" />
+```
+
+Turbo expects `GET /resources/3` to respond with some HTML that will look like this:
+
+```html
+<turbo-frame id="my-frame">
+  <span>This is resource 3</span>
+</turbo-frame>
+```
+
+Then it will swap the original frame with that payload, and the final result is:
+
+```html
+<!-- some HTML before the frame, for example: -->
+<img src="/hello.jpg" alt="hello" />
+
+<turbo-frame id="my-frame">
+  <span>This is resource 3</span>
+</turbo-frame>
+
+<!-- more HTML here, for example: -->
+<img src="/bye.jpg" alt="bye" />
+```
+
+If the rendered HTML does not contain `<turbo-frame id="my-frame">[...]</turbo-frame>` you will get the infamous `Content missing` error.
+
+So if the `GET /resources/3` request returns a full HTML page - which is what your app was doing before you added Turbo, so that's still probably what it does after you added it - then you want to break out of the frame. There are several ways to do it. For example, you can have a [global meta tag](https://turbo.hotwired.dev/handbook/frames#%E2%80%9Cbreaking-out%E2%80%9D-from-a-frame). That's not very convenient if you wanna be more granular.
+
+You can _try_ to have a top `target="_top"` in the frame like this:
+
+```html
+<turbo-frame id="my-frame" target="_top">
+  <a href="/resources/3">Show resource</a>
+</turbo-frame>
+```
+
+It may work in some cases, but I've been bitten by it, so I wouldn't recommend. Details are unimportant but let's say that these 2 snippets do **NOT** have the exact same behavior:
+
+```html
+<turbo-frame id="my-frame" target="_top">
+  <a href="/resources/3">Show resource</a>
+</turbo-frame>
+```
+is not the same as
+```html
+<turbo-frame id="my-frame">
+  <a href="/resources/3" data-turbo-frame="_top">Show resource</a>
+</turbo-frame>
+```
+
+because Turbo Streams responses are impacted by the `target="_top"` attribute.
```

Comment:
mm so maybe I'm wrong here, here is what I mean:

I have a page like `GET /delivery` which renders this html:

```html.erb
<%= turbo_frame_tag "delivery" do %>
  <%= form_tag update_delivery_path, method: :patch, data: {turbo_stream: true} do %>
    <!-- my form here, with autosubmit on radio change -->
    <a href="/whatever" data-turbo-frame="_top">A link that goes outside of the "delivery" frame</a>
  <% end %>
<% end %>
```
and I have a controller with 2 endpoints, `GET /delivery` and `PATCH /update_delivery`:

```ruby
def delivery
  # renders the above html
end

def update_delivery
  # update
  render :delivery # success
end
```

Both the link (to _top frame) and the form work:

![2024-09-25 17 15 32](https://github.com/user-attachments/assets/31a37b82-d5e5-448e-892d-967b670c6c46)

Now I'm gonna do the opposite, this is the html:

```diff
-<%= turbo_frame_tag "delivery" do %>
+<%= turbo_frame_tag "delivery", target: "_top" do %>
  <%= form_tag update_delivery_path, method: :patch, data: {turbo_stream: true} do %>
    <!-- my form here, with autosubmit on radio change -->
-    <a href="/whatever" data-turbo-frame="_top">A link that goes outside of the "delivery" frame</a>
+    <a href="/whatever">A link that goes outside of the "delivery" frame</a>
  <% end %>
<% end %>
```

the link works, but not the form:

![2024-09-25 17 19 03](https://github.com/user-attachments/assets/4d918775-2901-479c-8ee5-3b03cfce2a31)

the request is made correctly, but the response (`render :delivery`) will target the `_top` frame now. So I need to also update my form:

```diff
-<%= turbo_frame_tag "delivery" do %>
+<%= turbo_frame_tag "delivery", target: "_top" do %>
-  <%= form_tag update_delivery_path, method: :patch, data: {turbo_stream: true} do %>
+  <%= form_tag update_delivery_path, method: :patch, data: {turbo_stream: true, turbo_frame: "delivery"} do %>
    <!-- my form here, with autosubmit on radio change -->
-    <a href="/whatever" data-turbo-frame="_top">A link that goes outside of the "delivery" frame</a>
+    <a href="/whatever">A link that goes outside of the "delivery" frame</a>
  <% end %>
<% end %>
```

Reply:
I should rephrase this whole paragraph to give the form/stream example maybe 🤔

Reply:
I did rewrite that part to be more specific about turbo stream actions

Reply:
I don't follow, where are turbo streams in that example? 

The text you went with in the end is confusing me, it says now:
> Turbo Streams request made from within the frame, they also target the _top frame by default. 

A turbo stream response does not target anything. It's just turbo stream actions being executed. Unless you mean "HTML response with turbo stream tags inside the html"?

Reply:
@radanskoric I'm not sure how else to explain it. In this example:

template for `delivery.html.erb`

```html
<%= turbo_frame_tag "delivery" do %>
  <%= form_tag update_delivery_path, method: :patch, data: {turbo_stream: true} do %>
  <% end %>
<% end %>
```

controller:

```ruby
def delivery
  # renders template above
end

def update_delivery
  # do stuff
  render :delivery
end
```

The form submission is a turbo stream request to PATCH /update_delivery. The response of that request, by default, will target the `"delivery"` frame. because the form is in it. Which is what I want.

So if you change the frame itself to `target="_top"` then the response of the form submission now targets `_top`. which is not what I want. I only want links to target _top.

Reply:
That part is clear but this has nothing to do with Turbo Streams. There's no a single turbo stream tag in sight anywhere here. 

Which is why I was confused when you said: "because Turbo Streams responses are impacted by the `target="_top"` attribute." 

When a turbo stream response is returned, the target attribute should be irrelevant. And in the sample code you're not returning a turbo stream response. Did you mean to say: "Because **form submissions** are impacted by the target="_top" attribute?"

Reply:
ah ha ok, so my (new) phrasing is:

> if you have any Turbo Streams request made from within the frame, they also target the _top frame by default

So to me that form submission is a Turbo Streams request. It's not? (My original phrasing was confusing I agree)

Reply:
Ah, ok, I see now what you mean and what the confusion was. I don't call it a turbo streams request. I guess you can technically say it is because it will accept a **turbo streams response** but it doesn't have to (i.e. the accept header has both turbo stream and html listed). 

I just call it a form submission. I thought you were talking about a turbo streams **response** and that doesn't target a frame. Any turbo stream tag will target any element by id (or css) globally on the page, completely ignoring frames.

Reply:
Yeah ok, that makes sense